exporters/common/src/main/java/io/opentelemetry/exporter/internal/marshal/ProtoSerializer.java

```diff
@@ -187,8 +187,12 @@ public void writeSerializedMessage(byte[] protoSerialized, String jsonSerialized
 
   @Override
   public void close() throws IOException {
-    output.flush();
-    idCache.clear();
+    try {
+      output.flush();
+      idCache.clear();
+    } catch (IOException e) {
+      throw new IOException(e);
```

Comment:
I'm sorry for my ignorance - how exactly does this code result in `IllegalArgumentException` and how wrapping it in a try / catch solve it?

Reply:
> I'm sorry for my ignorance - how exactly does this code result in `IllegalArgumentException` and how wrapping it in a try / catch solve it?

public abstract class Marshaler {
  ...
  public final void writeBinaryTo(OutputStream output) throws IOException {
    try (Serializer serializer = new ProtoSerializer(output)) {
      writeTo(serializer);
    }
  }

  protected abstract void writeTo(Serializer output) throws IOException;
  ...
}

in Marshaler.writeBinaryTo() , an exception may be thrown, and ProtoSerializer implements AutoCloseable, so ProtoSerializer.close() will be executed, but the same error is thrown in this method, same error instance, then Throwable.addSuppressed() will be executed, that's what caused this problem

    Throwable.java
    public final synchronized void addSuppressed(Throwable exception) {
        if (exception == this) {
            throw new IllegalArgumentException("Self-suppression not permitted", exception);
        } else if (exception == null) {
            throw new NullPointerException("Cannot suppress a null exception.");
        } else if (this.suppressedExceptions != null) {
            if (this.suppressedExceptions == SUPPRESSED_SENTINEL) {
                this.suppressedExceptions = new ArrayList(1);
            }
            this.suppressedExceptions.add(exception);
        }
    }

by try/catch with ProtoSerializer.close() and throw a new IOException instance might solve the problem
It worked in my test sample